src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs

```diff
@@ -193,6 +190,64 @@ bool IHttpBodyControlFeature.AllowSynchronousIO
             }
         }
 
+        PipeWriter IResponseBodyPipeFeature.ResponseBodyPipe
+        {
+            get
+            {
+                if (_userSetPipeWriter != null)
+                {
+                    return _userSetPipeWriter;
+                }
+
+                if (!object.ReferenceEquals(_cachedResponseBodyStream, ResponseBody))
+                {
+                    var responsePipeWriter = new StreamPipeWriter(ResponseBody);
+                    ResponsePipeWriter = responsePipeWriter;
+                    _cachedResponseBodyStream = ResponseBody;
+                    OnCompleted((rpw) =>
+                    {
+                        ((StreamPipeWriter)rpw).Dispose();
+                        return Task.CompletedTask;
+                    }, responsePipeWriter);
+
+                }
+
+                return ResponsePipeWriter;
+            }
+            set
+            {
+                _userSetPipeWriter = value ?? throw new ArgumentNullException(nameof(value));
+                ResponsePipeWriter = _userSetPipeWriter;
+            }
+        }
+
+
+        Stream IHttpResponseFeature.Body
+        {
+            get
+            {
+                if (_userSetResponseBody != null)
+                {
+                    return _userSetResponseBody;
```

Comment:
I do find it kinda weird how any attempt to set the property, even back to it's original value, breaks the automatic stream-to-pipe-to-stream adaptation. For example.

```C#
var originalBody = context.Response.Body;
var ms = new MemoryStream()
try
{
    context.Response.Body = new MemoryStream();
    await next();
    ms.Position = 0;
    await ms.CopyToAsync(originalBody);
}
finally
{
    context.Response.Body = originalBody;
}

// Later...

context.Response.BodyPipe = new FooPipeWriter(context.Response.BodyPipe);

// Even later... The following uses the last user-set response stream,
// "originalBody", not an adapted FooPipeWriter :(

await context.Response.Body.WriteAsync(data, 0, data.Length); 
```=

Reply:
Bleh... yeah you are right. That isn't great behavior (though very nested). The problem I'm trying to avoid is when the user sets both the response body and response pipe in the same middleware, for example:

```C#
context.Response.Body = new MemoryStream();
context.Response.BodyPipe = new FooPipeWriter();
// Body is now a StreamPipeWriter
```

I don't see a great way to solve both of these problems.

Reply:
I'm not sure there is a great way to solve both problems. Preferring the user-set stream/pipe is important for the reason you highlighted, but it causes the kind of issue I highlighted.

That's why I'd prefer no magical getter/setter. It's way too hard to reason about.

Reply:
Let's chat about this tomorrow. I think I prefer your suggestion more than mine, but let's make sure we don't miss anything.

Reply:
I'd also say this is a non-blocker for getting this PR in.

Reply:
Based on offline feedback, this will always wrap whatever was set last.